README.md
# k8sprom-patch-controller

[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/5862/badge)](https://bestpractices.coreinfrastructure.org/projects/5862)
[![e2e](https://github.com/DoodleScheduling/k8sprom-patch-controller/workflows/e2e/badge.svg)](https://github.com/DoodleScheduling/k8sprom-patch-controller/actions)
[![report](https://goreportcard.com/badge/github.com/DoodleScheduling/k8sprom-patch-controller)](https://goreportcard.com/report/github.com/DoodleScheduling/k8sprom-patch-controller)
[![license](https://img.shields.io/github/license/DoodleScheduling/k8sprom-patch-controller.svg)](https://github.com/DoodleScheduling/k8sprom-patch-controller/blob/master/LICENSE)
[![release](https://img.shields.io/github/release/DoodleScheduling/k8sprom-patch-controller/all.svg)](https://github.com/DoodleScheduling/k8sprom-patch-controller/releases)

Apply patches to selected kubernetes resources based on prometheus queries.

## Example

```yaml
apiVersion: metrics.infra.doodle.com/v1beta1
kind: PrometheusPatchRule
metadata:
  name: annotate-namespace
spec:
  prometheus:
    address: http://prometheus-server.prometheus
  expr: |
    rate(nginx_ingress_controller_requests{exported_namespace="default"}[5m]) == 0
  for: 5m
  interval: 2m
  suspend: false
  json6902Patches:
  - target:
      version: v1
      resource: namespaces
      name: default
    patch:
    - op: add
      path: /metadata/annotations/has-ingress-traffic"
      value: "false"
```

## Details

### Prometheus expression
As soon as the given rule spec.expr evaluates to `true` the patches spec.patches get applied to the defined target `spec.patches[].target`.

### Pending state
You may define a window spec.for for which the rule will be in a pending condition similar to prometheus alerting rules.
As soon as the expression was `true` for the specified duration the patches get applied.

### Patches
Define a list of patches which needs a target selector as well as a list of JSON 6902 patch operations.
The target select requires at least the api version `version` as well as the resource group `resource` which is usually the kind in plural lowercase.

```
json6902Patches:
- target:
    version: v1
    resource: namespaces
    name: default
  patch:
  - op: add
    path: /metadata/annotations/has-ingress-traffic"
    value: "false"
```
Instead selecting a single resource you may also select multiple ones by left out the name field.
You can filter multiple onse by specifying a comma separated label select: `labelSelector: label=value,label2=value`.

### Interval
Defines in what interval the rule is evaluated.

### Suspend
The PrometheusPatchRule may be suspended setting spec.suspend to `true`. A suspended rule does not get reconciled, meaning no patches will be applied as long as the rule is suspended.

### Remove patches
By design patches are **not** removed if the defined expression evaluates to `false` and if the patches have been added before.
The way to achieve this is to crate another PrometheusPatchRule which expression does the opposite as well as opposite patches.

## Installation

### Requirements
You need a running prometheus [prometheus](https://prometheus.io/) (or any compatible prometheus service like [thanos](https://thanos.io/)).

### Permission
By default both the helm chart and the kustomize default base have a cluster rolebinding to cluster-admin.
Meaning the controller is granted full admin permission on the cluster.
This is needed as patch rules can target any kind of resources.
You may disable the binding and define fine grained cluster roles accordingly.

### Helm

Please see [chart/k8sprom-patch-controller](https://github.com/DoodleScheduling/k8sprom-patch-controller/tree/master/chart/k8sprom-patch-controller) for the helm chart docs.

### Manifests/kustomize

Alternatively you may get the bundled manifests in each release to deploy it using kustomize or use them directly.

## Configure the controller

You may change base settings for the controller using env variables (or alternatively command line arguments).
Available env variables:

| Name  | Description | Default |
|-------|-------------| --------|
| `METRICS_ADDR` | The address of the metric endpoint binds to. | `:9556` |
| `PROBE_ADDR` | The address of the probe endpoints binds to. | `:9557` |
| `ENABLE_LEADER_ELECTION` | Enable leader election for controller manager. | `false` |
| `LEADER_ELECTION_NAMESPACE` | Change the leader election namespace. This is by default the same where the controller is deployed. | `` |
| `NAMESPACES` | The controller listens by default for all namespaces. This may be limited to a comma delimited list of dedicated namespaces. | `` |
| `CONCURRENT` | The number of concurrent reconcile workers.  | `2` |
| `FIELD_MANAGER` | The name of the field manager used for server side apply https://kubernetes.io/docs/reference/using-api/server-side-apply/. | `k8sprom-patch-controller` |


CHANGELOG.md
# Changelog

## 0.1.2

**Release date:** 2022-04-01

### Bugfixes
* Fix env binding FIELD_MANAGER to cmd arg --field-manager


### Changes
* Change leader election id to k8sprom-patch-controller

## 0.1.1

**Release date:** 2022-03-31

### Bugfixes
* Fix infinite reconcile loop if patch failed and a for window is configured


## 0.1.0

**Release date:** 2022-03-29

Initial release


scripts/e2e/test.sh
#!/bin/bash
kubectl -n default apply -f config/testdata/rule.yaml
kubectl -n default wait prometheuspatchrule/annotate-namespace --for=condition=Active --timeout=1m
kubectl -n default wait prometheuspatchrule/annotate-namespace --for=condition=PatchApplied --timeout=15s
test $(kubectl get ns default -o=jsonpath='{.metadata.annotations.foo}') == "bar"


controllers/suite_test.go
/*
Copyright 2022 Doodle.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package controllers

import (
	"context"
	"fmt"
	"math/rand"
	"path/filepath"
	"testing"

	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	corev1 "k8s.io/api/core/v1"
	"k8s.io/client-go/dynamic"
	"k8s.io/client-go/kubernetes/scheme"
	"k8s.io/client-go/rest"
	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/client"
	"sigs.k8s.io/controller-runtime/pkg/envtest"
	logf "sigs.k8s.io/controller-runtime/pkg/log"
	"sigs.k8s.io/controller-runtime/pkg/log/zap"

	metricsinfradoodlecomv1beta1 "github.com/doodlescheduling/k8sprom-patch-controller/api/v1beta1"
	//+kubebuilder:scaffold:imports
)

// These tests use Ginkgo (BDD-style Go testing framework). Refer to
// http://onsi.github.io/ginkgo/ to learn more about Ginkgo.

var cfg *rest.Config
var k8sClient client.Client
var testEnv *envtest.Environment
var k8sManager ctrl.Manager
var ctx context.Context
var cancel context.CancelFunc

func TestAPIs(t *testing.T) {
	RegisterFailHandler(Fail)

	RunSpecs(t, "Controller Suite")
}

var _ = BeforeSuite(func() {
	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))

	By("bootstrapping test environment")
	testEnv = &envtest.Environment{
		CRDDirectoryPaths:     []string{filepath.Join("..", "config", "crd", "bases")},
		ErrorIfCRDPathMissing: false,
	}

	cfg, err := testEnv.Start()
	Expect(err).NotTo(HaveOccurred())
	Expect(cfg).NotTo(BeNil())

	err = corev1.AddToScheme(scheme.Scheme)
	Expect(err).NotTo(HaveOccurred())

	err = metricsinfradoodlecomv1beta1.AddToScheme(scheme.Scheme)
	Expect(err).NotTo(HaveOccurred())

	k8sManager, err = ctrl.NewManager(cfg, ctrl.Options{
		Scheme: scheme.Scheme,
	})
	Expect(err).ToNot(HaveOccurred())

	dynClient, err := dynamic.NewForConfig(k8sManager.GetConfig())
	Expect(err).ToNot(HaveOccurred(), "failed to setup dynClient")

	//+kubebuilder:scaffold:scheme
	// PrometheusPatchRule setup
	fmt.Printf("setup..................................")
	err = (&PrometheusPatchRuleReconciler{
		Client:       k8sManager.GetClient(),
		DynClient:    dynClient,
		FieldManager: "test-suite",
		Log:          ctrl.Log.WithName("controllers").WithName("PrometheusPatchRule"),
		Scheme:       k8sManager.GetScheme(),
		Recorder:     k8sManager.GetEventRecorderFor("PrometheusPatchRule"),
	}).SetupWithManager(k8sManager, PrometheusPatchRuleReconcilerOptions{MaxConcurrentReconciles: 10})

	Expect(err).ToNot(HaveOccurred(), "failed to setup PrometheusPatchRule")

	ctx, cancel = context.WithCancel(context.TODO())
	go func() {
		err = k8sManager.Start(ctx)
		Expect(err).ToNot(HaveOccurred())
	}()

	k8sClient, err = client.New(cfg, client.Options{Scheme: scheme.Scheme})
	Expect(err).NotTo(HaveOccurred())
	Expect(k8sClient).NotTo(BeNil())

})

var _ = AfterSuite(func() {
	cancel()
	By("tearing down the test environment")
	err := testEnv.Stop()
	Expect(err).ToNot(HaveOccurred())
})

var letterRunes = []rune("abcdefghijklmnopqrstuvwxyz1234567890")

func randStringRunes(n int) string {
	b := make([]rune, n)
	for i := range b {
		b[i] = letterRunes[rand.Intn(len(letterRunes))]
	}
	return string(b)
}


api/v1beta1/patchrule_types.go
/*
Copyright 2022 Doodle.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package v1beta1

import (
	extv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
	apimeta "k8s.io/apimachinery/pkg/api/meta"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

const (
	ActiveCondition             = "Active"
	FailedReason                = "Failed"
	InactiveReason              = "Inactive"
	PendingReason               = "Pending"
	ActiveReason                = "Active"
	InvalidPrometheusURLReason  = "InvalidPrometheusURL"
	PrometheusQueryFailedReason = "PrometheusQueryFailed"
	PatchAppliedCondition       = "PatchApplied"
	PatchApplyFailedReason      = "Failed"
	PatchAppliedReason          = "Applied"
	NoPatchFoundReason          = "NoPatchFound"
)

// PrometheusPatchRuleSpec defines the desired state of PrometheusPatchRule
type PrometheusPatchRuleSpec struct {
	// Prometheus holds information about where to find prometheus
	// +required
	Prometheus PrometheusSpec `json:"prometheus"`

	// Interval is the duration in which the expression gets evaluated
	// +required
	Interval metav1.Duration `json:"interval,omitempty"`

	// Expression is the prometheus .query
	// +required
	Expr string `json:"expr,omitempty"`

	// For is a durstion for how long the rule should be in pending before apply patches.
	// +required
	For metav1.Duration `json:"for,omitempty"`

	// .JSON6902Patches define to what target are applied what patches
	// +required
	JSON6902Patches []JSON6902Patch `json:"json6902Patches,omitempty"`

	// Suspend may suspend reconciliation of the resource.
	// +optional
	Suspend bool `json:"suspend,omitempty"`
}

// PrometheusSpec contains specs for accessing prometheus
type PrometheusSpec struct {
	Address string `json:"address"`
}

// JSON6902Patch is a target selector and a list of JSON6902 patches
type JSON6902Patch struct {
	// Patch contains JSON6902 patches with
	// an array of operation objects.
	// +required
	Patch []JSONPatch `json:"patch,omitempty"`

	// Target points to the resources that the patch document should be applied to.
	// +optional
	Target Selector `json:"target,omitempty"`
}

// JSONPatch is a JSON 6902 conform patch
type JSONPatch struct {
	OP    string     `json:"op"`
	Path  string     `json:"path"`
	Value extv1.JSON `json:"value"`
}

// Selector specifies a set of resources. Any resource that matches intersection of all conditions is included in this
// set.
type Selector struct {
	// Group is the API group to select resources from.
	// Together with Version and Kind it is capable of unambiguously identifying and/or selecting resources.
	// https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/api-group.md
	// +optional
	Group string `json:"group,omitempty"`

	// Version of the API Group to select resources from.
	// Together with Group and Kind it is capable of unambiguously identifying and/or selecting resources.
	// https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/api-group.md
	// +optional
	Version string `json:"version,omitempty"`

	// Kind of the API Group to select resources from.
	// Together with Group and Version it is capable of unambiguously
	// identifying and/or selecting resources.
	// https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/api-group.md
	// +optional
	Resource string `json:"resource,omitempty"`

	// Namespace to select resources from.
	// +optional
	Namespace string `json:"namespace,omitempty"`

	// Name to match resources with.
	// +optional
	Name string `json:"name,omitempty"`

	// LabelSelector is a string that follows the label selection expression
	// https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#api
	// It matches with the resource labels.
	// +optional
	LabelSelector string `json:"labelSelector,omitempty"`
}

// PrometheusPatchRuleStatus defines the observed state of PrometheusPatchRule
type PrometheusPatchRuleStatus struct {
	// Conditions holds the conditions for the PrometheusPatchRule.
	// +optional
	Conditions []metav1.Condition `json:"conditions,omitempty"`
}

// ConditionalResource is a resource with conditions
type conditionalResource interface {
	GetStatusConditions() *[]metav1.Condition
	GetGeneration() int64
}

// setResourceCondition sets the given condition with the given status,
// reason and message on a resource.
func setResourceCondition(resource conditionalResource, condition string, status metav1.ConditionStatus, reason, message string) {
	conditions := resource.GetStatusConditions()

	newCondition := metav1.Condition{
		Type:               condition,
		Status:             status,
		Reason:             reason,
		Message:            message,
		ObservedGeneration: resource.GetGeneration(),
	}

	apimeta.SetStatusCondition(conditions, newCondition)
}

// PrometheusPatchRuleNotActive
func PrometheusPatchRuleNotActive(rule PrometheusPatchRule, reason, message string) PrometheusPatchRule {
	setResourceCondition(&rule, ActiveCondition, metav1.ConditionFalse, reason, message)
	return rule
}

// PrometheusPatchRuleActive
func PrometheusPatchRuleActive(rule PrometheusPatchRule, reason, message string) PrometheusPatchRule {
	setResourceCondition(&rule, ActiveCondition, metav1.ConditionTrue, reason, message)
	return rule
}

// PrometheusPatchRuleNoPatchApplied
func PrometheusPatchRuleNoPatchApplied(rule PrometheusPatchRule, reason, message string) PrometheusPatchRule {
	setResourceCondition(&rule, PatchAppliedCondition, metav1.ConditionFalse, reason, message)
	return rule
}

// PrometheusPatchRulePatchApplied
func PrometheusPatchRulePatchApplied(rule PrometheusPatchRule, reason string) PrometheusPatchRule {
	setResourceCondition(&rule, PatchAppliedCondition, metav1.ConditionTrue, reason, "")
	return rule
}

// GetStatusConditions returns a pointer to the Status.Conditions slice
func (in *PrometheusPatchRule) GetStatusConditions() *[]metav1.Condition {
	return &in.Status.Conditions
}

// +kubebuilder:object:root=true
// +kubebuilder:subresource:status
// +kubebuilder:printcolumn:name="Active",type="string",JSONPath=".status.conditions[?(@.type==\"Active\")].status",description=""
// +kubebuilder:printcolumn:name="Reason",type="string",JSONPath=".status.conditions[?(@.type==\"Active\")].reason",description=""
// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp",description=""

// PrometheusPatchRule is the Schema for the patchrules API
type PrometheusPatchRule struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec   PrometheusPatchRuleSpec   `json:"spec,omitempty"`
	Status PrometheusPatchRuleStatus `json:"status,omitempty"`
}

//+kubebuilder:object:root=true

// PrometheusPatchRuleList contains a list of PrometheusPatchRule
type PrometheusPatchRuleList struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ListMeta `json:"metadata,omitempty"`
	Items           []PrometheusPatchRule `json:"items"`
}

func init() {
	SchemeBuilder.Register(&PrometheusPatchRule{}, &PrometheusPatchRuleList{})
}


api/v1beta1/groupversion_info.go
/*
Copyright 2022 Doodle.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package v1beta1 contains API Schema definitions for the metrics.infra.doodle.com v1beta1 API group
//+kubebuilder:object:generate=true
//+groupName=metrics.infra.doodle.com
package v1beta1

import (
	"k8s.io/apimachinery/pkg/runtime/schema"
	"sigs.k8s.io/controller-runtime/pkg/scheme"
)

var (
	// GroupVersion is group version used to register these objects
	GroupVersion = schema.GroupVersion{Group: "metrics.infra.doodle.com", Version: "v1beta1"}

	// SchemeBuilder is used to add go types to the GroupVersionKind scheme
	SchemeBuilder = &scheme.Builder{GroupVersion: GroupVersion}

	// AddToScheme adds the types in this group-version to the given scheme.
	AddToScheme = SchemeBuilder.AddToScheme
)


api/v1beta1/zz_generated.deepcopy.go
//go:build !ignore_autogenerated
// +build !ignore_autogenerated

/*
Copyright 2022 Doodle.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Code generated by controller-gen. DO NOT EDIT.

package v1beta1

import (
	"k8s.io/apimachinery/pkg/apis/meta/v1"
	runtime "k8s.io/apimachinery/pkg/runtime"
)

// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
func (in *JSON6902Patch) DeepCopyInto(out *JSON6902Patch) {
	*out = *in
	if in.Patch != nil {
		in, out := &in.Patch, &out.Patch
		*out = make([]JSONPatch, len(*in))
		for i := range *in {
			(*in)[i].DeepCopyInto(&(*out)[i])
		}
	}
	out.Target = in.Target
}

// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new JSON6902Patch.
func (in *JSON6902Patch) DeepCopy() *JSON6902Patch {
	if in == nil {
		return nil
	}
	out := new(JSON6902Patch)
	in.DeepCopyInto(out)
	return out
}

// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
func (in *JSONPatch) DeepCopyInto(out *JSONPatch) {
	*out = *in
	in.Value.DeepCopyInto(&out.Value)
}

// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new JSONPatch.
func (in *JSONPatch) DeepCopy() *JSONPatch {
	if in == nil {
		return nil
	}
	out := new(JSONPatch)
	in.DeepCopyInto(out)
	return out
}

// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
func (in *PrometheusPatchRule) DeepCopyInto(out *PrometheusPatchRule) {
	*out = *in
	out.TypeMeta = in.TypeMeta
	in.ObjectMeta.DeepCopyInto(&out.ObjectMeta)
	in.Spec.DeepCopyInto(&out.Spec)
	in.Status.DeepCopyInto(&out.Status)
}

// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new PrometheusPatchRule.
func (in *PrometheusPatchRule) DeepCopy() *PrometheusPatchRule {
	if in == nil {
		return nil
	}
	out := new(PrometheusPatchRule)
	in.DeepCopyInto(out)
	return out
}

// DeepCopyObject is an autogenerated deepcopy function, copying the receiver, creating a new runtime.Object.
func (in *PrometheusPatchRule) DeepCopyObject() runtime.Object {
	if c := in.DeepCopy(); c != nil {
		return c
	}
	return nil
}

// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
func (in *PrometheusPatchRuleList) DeepCopyInto(out *PrometheusPatchRuleList) {
	*out = *in
	out.TypeMeta = in.TypeMeta
	in.ListMeta.DeepCopyInto(&out.ListMeta)
	if in.Items != nil {
		in, out := &in.Items, &out.Items
		*out = make([]PrometheusPatchRule, len(*in))
		for i := range *in {
			(*in)[i].DeepCopyInto(&(*out)[i])
		}
	}
}

// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new PrometheusPatchRuleList.
func (in *PrometheusPatchRuleList) DeepCopy() *PrometheusPatchRuleList {
	if in == nil {
		return nil
	}
	out := new(PrometheusPatchRuleList)
	in.DeepCopyInto(out)
	return out
}

// DeepCopyObject is an autogenerated deepcopy function, copying the receiver, creating a new runtime.Object.
func (in *PrometheusPatchRuleList) DeepCopyObject() runtime.Object {
	if c := in.DeepCopy(); c != nil {
		return c
	}
	return nil
}

// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
func (in *PrometheusPatchRuleSpec) DeepCopyInto(out *PrometheusPatchRuleSpec) {
	*out = *in
	out.Prometheus = in.Prometheus
	out.Interval = in.Interval
	out.For = in.For
	if in.JSON6902Patches != nil {
		in, out := &in.JSON6902Patches, &out.JSON6902Patches
		*out = make([]JSON6902Patch, len(*in))
		for i := range *in {
			(*in)[i].DeepCopyInto(&(*out)[i])
		}
	}
}

// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new PrometheusPatchRuleSpec.
func (in *PrometheusPatchRuleSpec) DeepCopy() *PrometheusPatchRuleSpec {
	if in == nil {
		return nil
	}
	out := new(PrometheusPatchRuleSpec)
	in.DeepCopyInto(out)
	return out
}

// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
func (in *PrometheusPatchRuleStatus) DeepCopyInto(out *PrometheusPatchRuleStatus) {
	*out = *in
	if in.Conditions != nil {
		in, out := &in.Conditions, &out.Conditions
		*out = make([]v1.Condition, len(*in))
		for i := range *in {
			(*in)[i].DeepCopyInto(&(*out)[i])
		}
	}
}

// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new PrometheusPatchRuleStatus.
func (in *PrometheusPatchRuleStatus) DeepCopy() *PrometheusPatchRuleStatus {
	if in == nil {
		return nil
	}
	out := new(PrometheusPatchRuleStatus)
	in.DeepCopyInto(out)
	return out
}

// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
func (in *PrometheusSpec) DeepCopyInto(out *PrometheusSpec) {
	*out = *in
}

// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new PrometheusSpec.
func (in *PrometheusSpec) DeepCopy() *PrometheusSpec {
	if in == nil {
		return nil
	}
	out := new(PrometheusSpec)
	in.DeepCopyInto(out)
	return out
}

// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
func (in *Selector) DeepCopyInto(out *Selector) {
	*out = *in
}

// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new Selector.
func (in *Selector) DeepCopy() *Selector {
	if in == nil {
		return nil
	}
	out := new(Selector)
	in.DeepCopyInto(out)
	return out
}


main.go
/*
Copyright 2022 Doodle.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"flag"
	"os"
	"strings"

	// Import all Kubernetes client auth plugins (e.g. Azure, GCP, OIDC, etc.)
	// to ensure that exec-entrypoint and run can make use of them.
	"k8s.io/client-go/dynamic"
	_ "k8s.io/client-go/plugin/pkg/client/auth"

	"github.com/spf13/pflag"
	"github.com/spf13/viper"
	"k8s.io/apimachinery/pkg/runtime"
	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/cache"
	"sigs.k8s.io/controller-runtime/pkg/healthz"
	"sigs.k8s.io/controller-runtime/pkg/log/zap"

	metricsinfradoodlecomv1beta1 "github.com/doodlescheduling/k8sprom-patch-controller/api/v1beta1"
	"github.com/doodlescheduling/k8sprom-patch-controller/controllers"
	//+kubebuilder:scaffold:imports
)

var (
	scheme   = runtime.NewScheme()
	setupLog = ctrl.Log.WithName("setup")
)

func init() {
	utilruntime.Must(clientgoscheme.AddToScheme(scheme))

	utilruntime.Must(metricsinfradoodlecomv1beta1.AddToScheme(scheme))
	//+kubebuilder:scaffold:scheme
}

var (
	metricsAddr             = ":9556"
	probesAddr              = ":9557"
	enableLeaderElection    = false
	leaderElectionNamespace string
	namespaces              = ""
	concurrent              = 2
	fieldManager            = "k8sprom-patch-controller"
)

func main() {
	flag.StringVar(&metricsAddr, "metrics-addr", metricsAddr, "The address of the metric endpoint binds to.")
	flag.StringVar(&probesAddr, "probe-addr", probesAddr, "The address of the probe endpoints bind to.")
	flag.StringVar(&fieldManager, "field-manager", fieldManager, "The name of the field maanger used for server side apply https://kubernetes.io/docs/reference/using-api/server-side-apply/.")
	flag.BoolVar(&enableLeaderElection, "enable-leader-election", enableLeaderElection,
		"Enable leader election for controller manager. "+
			"Enabling this will ensure there is only one active controller manager.")
	flag.StringVar(&leaderElectionNamespace, "leader-election-namespace", leaderElectionNamespace,
		"Specify a different leader election namespace. It will use the one where the controller is deployed by default.")
	flag.StringVar(&namespaces, "namespaces", namespaces,
		"The controller listens by default for all namespaces. This may be limited to a comma delimted list of dedicated namespaces.")
	flag.IntVar(&concurrent, "concurrent", concurrent,
		"The number of concurrent reconcile workers. By default this is 2.")

	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
	pflag.Parse()
	ctrl.SetLogger(zap.New(zap.UseDevMode(true)))

	// Import flags into viper and bind them to env vars
	// flags are converted to upper-case, - is replaced with _
	err := viper.BindPFlags(pflag.CommandLine)
	if err != nil {
		setupLog.Error(err, "Failed parsing command line arguments")
		os.Exit(1)
	}

	replacer := strings.NewReplacer("-", "_")
	viper.SetEnvKeyReplacer(replacer)
	viper.AutomaticEnv()

	opts := ctrl.Options{
		Scheme:                  scheme,
		MetricsBindAddress:      viper.GetString("metrics-addr"),
		HealthProbeBindAddress:  viper.GetString("probe-addr"),
		LeaderElection:          viper.GetBool("enable-leader-election"),
		LeaderElectionNamespace: viper.GetString("leader-election-namespace"),
		LeaderElectionID:        "k8sprom-patch-controller",
	}

	ns := strings.Split(viper.GetString("namespaces"), ",")
	if len(ns) > 0 && ns[0] != "" {
		opts.NewCache = cache.MultiNamespacedCacheBuilder(ns)
		setupLog.Info("watching dedicated namespaces", "namespaces", ns)
	} else {
		setupLog.Info("watching all namespaces")
	}

	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), opts)
	if err != nil {
		setupLog.Error(err, "unable to start manager")
		os.Exit(1)
	}

	dynClient, err := dynamic.NewForConfig(mgr.GetConfig())
	if err != nil {
		setupLog.Error(err, "unable to create dynamic client")
		os.Exit(1)
	}

	if err = (&controllers.PrometheusPatchRuleReconciler{
		Client:       mgr.GetClient(),
		DynClient:    dynClient,
		FieldManager: viper.GetString("field-manager"),
		Log:          ctrl.Log.WithName("controllers").WithName("PrometheusPatchRule"),
		Scheme:       mgr.GetScheme(),
		Recorder:     mgr.GetEventRecorderFor("PrometheusPatchRule"),
	}).SetupWithManager(mgr, controllers.PrometheusPatchRuleReconcilerOptions{MaxConcurrentReconciles: viper.GetInt("concurrent")}); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "PrometheusPatchRule")
		os.Exit(1)
	}
	//+kubebuilder:scaffold:builder

	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
		setupLog.Error(err, "unable to set up health check")
		os.Exit(1)
	}
	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
		setupLog.Error(err, "unable to set up ready check")
		os.Exit(1)
	}

	setupLog.Info("starting manager")
	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
		setupLog.Error(err, "problem running manager")
		os.Exit(1)
	}
}
